frontend/app/work_packages/controllers/details-tab-watchers-controller.js

```diff
@@ -26,127 +26,55 @@
-        // $scope.availableWatchers = data.embedded.elements;
+        data.embedded.elements.forEach(function(watcher) {
+          $scope.vm.availableWatchers.push(watcher);
+          if (existingWatcherIds.indexOf(getResourceIdentifier(watcher)) != -1) {
+            $scope.vm.selectedWatchers.push(watcher);
+          }
+        });
       });
-  }
-
-  $scope.getAvailableWatchers = function(term, result) {
-    var watchers = $scope.availableWatchers.map(function(watcher) {
-      return { id: watcher.props.id, name: watcher.props.name };
-    });
-
-    result($filter('filter')(watchers, {name: term}));
-  };
-
-  function addWatcher(newValue, oldValue) {
-    if (newValue) {
-      var id = newValue.id;
-
-      if (id) {
-        $scope.workPackage.link('addWatcher', {user_id: id})
-          .fetch({ajax: {method: 'POST'}})
-          .then(addWatcherSuccess, $scope.outputError);
-      }
-    }
-  }
-
-  function addWatcherSuccess() {
-    $scope.outputMessage(I18n.t("js.label_watcher_added_successfully"));
-    $scope.refreshWorkPackage();
-
-    $scope.watcher.selected = null;
+  });
 
-    $scope.focusElementIndex = ADD_WATCHER_SELECT_INDEX;
-  }
-
-  $scope.deleteWatcher = function(watcher) {
-    watcher.links.removeWatcher
-      .fetch({ ajax: watcher.links.removeWatcher.props })
-      .then(deleteWatcherSuccess(watcher), $scope.outputError);
-  };
-
-  function deleteWatcherSuccess(watcher) {
-    $scope.outputMessage(I18n.t("js.label_watcher_deleted_successfully"));
-    removeWatcherFromList(watcher);
-  }
-
-  function removeWatcherFromList(watcher) {
-    var index = $scope.watchers.indexOf(watcher);
-
-    if (index >= 0) {
-      $scope.watchers.splice(index, 1);
-
-      updateWatcherFocus(index);
-      $scope.$emit('workPackageRefreshRequired');
+  $scope.$watch('vm.selectedWatchers', function(newWatcherList, oldWatcherList) {
+    if (newWatcherList.length == oldWatcherList.length) {
+      // nothing happened
+      return;
     }
-  }
-
-  function updateWatcherFocus(index) {
-    if ($scope.watchers.length == 0) {
-      $scope.focusElementIndex = ADD_WATCHER_SELECT_INDEX;
+    if (newWatcherList.length > oldWatcherList.length) {
+      // add watcher
+      var newWatcher = _.difference(newWatcherList, oldWatcherList).pop();
+      $scope.workPackage.link('addWatcher', {'user_id': newWatcher.props.id})
+        .fetch({ajax: {method: 'POST'}});
     } else {
-      $scope.focusElementIndex = (index < $scope.watchers.length) ? index : $scope.watchers.length - 1;
+      // remove watcher
+      $scope.refreshWorkPackage(function() {
+        var doomedWatcher = _.difference(oldWatcherList, newWatcherList).pop();
+        var link = _.find($scope.workPackage.embedded.watchers, function(embeddedWatcher) {
+          return embeddedWatcher.props.id == doomedWatcher.props.id;
+        }).links.removeWatcher;
+        link.fetch({ ajax: link.props });
```

Comment:
@myabc The best I found was a discussion on hyperagent github about how they would like to keep hyperagent clean and if you wish, you can create something like hyperagent-forms.